feat: add the initial tests.md (#15)
PaloAltoNetworks/terraform-best-practices: tests.md

speak to the real cloud anymore.

Such large investment appears for now to be only possible through an organization-level decision, not a team-level one.
So the default is not to do the testing this way. Instead we resort to mainly `terraform apply` to the real cloud,
which one would normally call integration testing. Just as for example Go/Python projects' integration tests which could
speak to the third-party API or spin up the database instance.

As a corollary to this, the terms "Unit Test" and "Integration Test" should be used with care with Terraform. If you
mean something else than this section, clarify exactly what are your definitions and how the distinction matters in your
context.

The only test-like action which is designed to be quite quick with Terraform is the good old `terraform plan`. It's
worth running, but it can hardly be considered an exhaustive test. The Plan does not dynamically generate the
identifiers, instead reporting them as `id = (known after apply)`, so object-to-object relations cannot be verified
at all.

## Test Mainly An Empty Terraform State

> The assumption is that we mainly run tests starting from an empty tfstate. For most kinds of tests, maintaining

Reviewer: ```suggestion
The assumption is that we mainly run tests starting from an empty Terraform [state](https://www.terraform.io/docs/language/state/index.html) file. For most kinds of tests, maintaining
```

any permanent cloud objects and the associated Terraform state would be quite hard. The main reason is not the cost,

> but the fact that breaking changes in the code tend to destroy/recreate a lot of such cloud objects. Such destroys can

Reviewer: ```suggestion
but the fact that breaking changes in the code tend to destroy/recreate a lot of cloud resources. Such destroys can
```
nit: Tu not repeat "such" between two sentences?

quite often enter a conflict of some kind preventing any further Terraform modification to bring back the well-known
state for subsequent tests.

## Test Cases: Simple Or Complex?

> Per the previous sections, we resort to using real Apply in tests, and it usually brings in some cloud cost. (The Apply
> means the stage after the plan has been already shown to the user, approved, and when it is being deployed to the real
> world.)

Reviewer: ```suggestion
Per the previous sections, we resort to using real Apply in tests, and it usually brings in some cloud cost - the Apply
means the stage after the plan has been already shown to the user, approved, and when it is being deployed to the real
world.
```

> If quick and cheap unit tests were possible, it would be preferable to have each test case just cleanly one thing

Reviewer: ```suggestion
If quick and cheap unit tests were possible, it would be preferable to have each test case responsible for one thing
```
Nit: I'm not sure about the intention behind the sentence in the original version - this is how I understand what you meant? Ignore me if I'm wrong of course :)

and one thing only. Such code is usually readable, small, and clear. However, it remains an open question whether
to accept the associated cost of one more Apply. In some cases we prefer to sacrifice some readability and perform
multiple checks after a single Apply run. The reason is that the tests are run more often than they're read by a human.

But if you decide to create such complex test code (one Apply with multiple checks), structure it in a way that shows
that the checks are in fact independent. Use comments as needed to make that clear. This will reduce

> the ball-of-spaghetti flavor and rescue some readability.

Reviewer: ```suggestion
the ball-of-spaghetti flavor and rescue some readability. And your colleagues will not resent you as much as they used to.
```

Author: Ouch.

  u_false   = length(random_pet.this.id) < 0
  u_string  = random_pet.this.id
  u_int     = length(random_pet.this.id)
  u_cidr    = "10.0.${local.u_int}.0/24"
}
```

This way you can replace a `true` known bool value with `local.utrue` unknown bool value with ease. (The `local.utrue`
is always true, because the `length()` is always greater than zero.) This way it is possible to set every input
of the module to an unknown value in quite a readable way.

This pattern uses `random_pet`, a very simple Hashicorp-provided resource which simply generates names,
e.g. "relevant-horse". It is very fast, does not communicate to any cloud, and does not incur any cost, but otherwise
does not bring any real functionality here - the code above would be usable even if we would use `aws_vpc` instead.

### Outputs

The majority of module's outputs should be consumable by subsequent Terraform code. Only the minority of outputs
exist to be merely displayed to a user when Terraform completes. We can test for some very predictable scenarios of how
our module's output is consumed by some typical user-side code. It turns out that unknown values impact this too.

> The chief scenario is when an output is a `map`. There is not much one can do with a map, except to perform a series

Reviewer: Nit: "main" scenario instead of "chief"?

> # Testing of Terraform Modules Explained

Reviewer 2: Do you have any thoughts on how to monitor the percentage of test coverage metric?

Author: Ah no, not really.

## Are Unit Tests Viable?

A useful definition of a unit test is one which mainly spends its run time in the code that we own, and very little
time in the code owned by other organizations or teams. There are other definitions, but the usefulness of this
is dictated by tightening the loop of developer's work. The progress is faster with a five-seconds test run than with
a five-hours test run as the developer waits for the result to decide whether to go further with the work or to go
back and correct the last portion of their work. The more time the test run uses for the others' code, the less time

> it spends inside our code, to catch the kind of errors *this team* is able to fix by ourselves.

Reviewer 2: you may want to scope the unit tests down to the method/function level. 

A five hour test is still valuable from the stand point of how that method interacts with the rest of the code base, and should run less often.

Author: Yes. (Does the paragraph need corrections/modification/deletion?)

example may be 100 seconds versus less than 1 second.

To enable running multiple test cases in much shorter time than that, one can either mock the provider
or mock the cloud's REST API. Such a "mocked" component basically either returns hardcoded results or an error
when anything unexpected happens, so it only takes microseconds or so. However such method requires either:

- the Terraform team to be versed in Go language, to be able to implement their own entire mock provider,
- or, alternatively, the Terraform team to be versed enough in some language to implement their own cloud REST API,
to have intimate knowledge of the original REST API of that cloud, and to be able to configure the provider to not to
speak to the real cloud anymore.

Such large investment appears for now to be only possible through an organization-level decision, not a team-level one.
So the default is not to do the testing this way. Instead we resort to mainly `terraform apply` to the real cloud,
which one would normally call integration testing. Just as for example Go/Python projects' integration tests which could
speak to the third-party API or spin up the database instance.

> As a corollary to this, the terms "Unit Test" and "Integration Test" should be used with care with Terraform. If you

Reviewer 2: https://www.hashicorp.com/blog/testing-hashicorp-terraform

Author: Please clarify? Do you mean to use that blog entry as a definite reference and modify this section to be less speculative in tone and a more firm? ("howto" mode)

Author: I've removed the speculative bits.

# Testing of Terraform Modules Explained

> ## Unit Tests Do Not Use Apply

Author: Changed the title. The titles are used as url fragments, so they better be stable. Please double-check all titles.

Reviewer: Checked, consistent.

  u_false   = length(random_pet.this.id) < 0
  u_string  = random_pet.this.id
  u_number  = length(random_pet.this.id)
  u_cidr    = "10.0.${local.u_number}.0/24"
}
```

This way you can replace a `true` known bool value with `local.u_true` unknown bool value with ease. (The `local.u_true`
is always equal to `true`, because the `length()` is always greater than zero. But only the Apply stage determines
that, the Plan stage assumes it can be either true or false.) Same for other types of inputs: you can replace any string
by `local.u_string`, and number by `local.u_number`, and so on. This way it is possible to set every input of the module
to an unknown value in quite a readable way.

This pattern uses `random_pet`, a very simple Hashicorp-provided resource which simply generates names,
e.g. "relevant-horse". It is very fast, does not communicate to any cloud, and does not incur any cost, but otherwise
does not bring any real functionality here - the code above would be usable even if we would use `aws_vpc` instead.

### Outputs

The majority of module's outputs should be consumable by subsequent Terraform code. Only the minority of outputs

> exist to be merely displayed to a user when Terraform completes. We can test for some very predictable scenarios of how

Reviewer: ```suggestion
exist to be merely displayed to a user when a Terraform run completes. We can test for some very predictable scenarios of how
```
Nit

our module's output is consumed by some typical user-side code. It turns out that unknown values impact this too.

The main scenario is when an output is a `map`. There is not much one can do with a map, except to perform a series
of transformations and finally feed it to some `for_each` block at some point. But it's just impossible if the keys
of the map are an unknown set - the Plan will just fail with "Invalid for_each argument". The module theoretically works

> but prevents any Lego®-style building. (Such a map output can still be displayed to users however.)

Reviewer: ```suggestion
but prevents any Lego®-style building (such a map output can still be displayed to users however).
```

```hcl2
module "mymodule" {
  source = "../../modules/mymodule"

  /* inputs */
}

### Feed all the map outputs to a dummy for_each ###

resource "random_pet" "consume_mymodule_my_map_output" {
  for_each = module.mymodule.my_map_output
}

resource "random_pet" "consume_mymodule_second" {
  for_each = module.mymodule.second
}

resource "random_pet" "consume_mymodule_third" {
  for_each = module.mymodule.third
}
```

On Terraform 0.13 and above condense that to a single block:

```hcl2
resource "random_pet" "consume_maps" {
  for_each = merge(
    module.mymodule.my_map_output,
    module.mymodule.second,
    module.mymodule.third,

    module.mymodule_alt.my_map_output,
    module.mymodule_alt.second,
    module.mymodule_alt.third,
  )
}
```

The `_alt` above refers to an alternate call to the same module, with a different set of known inputs. For example,
if the `vpc` module has the input `create_vpc` which is true by default, an alternate call would have it false.

> This would test Plan on multiple execution paths within the same code.

Author: Am I going overboard? Split it to another doc?

Reviewer: IMO - keep it this way, I prefer having this in one bigger doc, than searching through a few smaller ones.

Author: 👍 Will do so. 👍

> Similarly, sometimes an output is a list, but a user can still be predicted to convert it to a map anyway. Lists

Reviewer: Not sure if I understand this line - do you mean that the user can predict the list values and convert them in to a map?

Author: Let's say a module.mymodule.subnet_ids is a list of strings `["subnet1", "subnet2"]`. You cannot feed a list to a for_each, Terraform never accepts that. But on the other hand, sometimes you just know that some module's users will try to do something like this:

```
resource "aws_route_table_association" "this" {
  for_each = { for v in module.mymodule.subnet_names: v => v }

  subnet_id = each.value
}
```

Better test that list against that future, if you can predict it. (My example output would fail such test and would need to be fixed/restructured.)

Ok, how can I make this line of doc better then?

Reviewer: "You cannot feed a list to a for_each, Terraform never accepts that." - yeaaah but you can always convert a list to a set for strings :) 

IMO changing line 179 to something like that and adding the explanation you just provided to me would be beneficial.
`Similarly, sometimes an output is a list, but a user can still insist it to convert it to a map anyway.`

Author: I've modified, please re-review that part.

I don't want to even mention `toset()`, because it doesn't solve the problem.

I cannot see it myself when only looking at the code, but it only proves how beneficial this type of testing is. It will just fail and stop me:

```txt
│ Error: Invalid for_each argument
│ 
│   on main.tf line 331, in resource "random_pet" "set":
│  331:   for_each = toset([local.u_string, "subnet2"])
│     ├────────────────
│     │ local.u_string is a string, known only after apply
│ 
│ The "for_each" value depends on resource attributes that cannot be
│ determined until apply, so Terraform cannot predict how many instances will
│ be created. To work around this, use the -target argument to first apply
│ only the resources that the for_each depends on.
╵}
```

Author: Please re-review @damianfedeczko 😄

however are more often used as resource arguments in practice, which are immune to unknown values. For example
`aws_instance.vpc_security_group_ids` expects a list of strings and works fine whether the list is known or unknown, so
it has completely different behavior than `for_each`. (Map arguments would be immune for the same reason, but they are

> rarely seen in providers except for the ubiquitous `tags` maps.) How to consume a list:

Author: New thought. Does it need more explanation or less?

Reviewer: I want to check if I understand/l
`{ for v in module.mymodule.my_list : v => true }`
Would the result of the expression be something like the output shown below, if we assume that `module.mymodule.my_list` is for example a list with the following elements `[foo, bar, damian]`:
```terraform
{
  foo = True
  bar = True
  damian = True
}
```

Author: Yes, that's the result if the list is good (yours is). If the list is not consumable, such as `[local.u_string, "damian"]`, the result will be `Error: Invalid for_each`.

> it has completely different behavior than `for_each`. (Map arguments would be immune for the same reason, but they are
> rarely seen in providers except for the ubiquitous `tags` maps.) How to consume a list:

Reviewer: ```suggestion
it has completely different behavior than `for_each` (map arguments would be immune for the same reason, but they are
rarely seen in providers except for the ubiquitous `tags` maps). 

How to consume a list:
```
